File: backend/storage-service/src/main/java/com/example/storageservice/kafka/producer/KafkaProducerService.java
package com.example.storageservice.kafka.producer;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;

@Slf4j
@Service
@RequiredArgsConstructor
public class KafkaProducerService {

    private final KafkaTemplate<String, Object> kafkaTemplate;
    private static final String TOPIC = "topic-member-rollback";

    public void sendMessage(String key, Object message) {
        kafkaTemplate.send(TOPIC, key, message)
                .whenComplete((result, ex) -> {
                    if (ex == null) {
                        log.info("Message sent successfully: {}", result.getRecordMetadata().topic() + " / " + message);
                    } else {
                        log.info("Message sent failed: {}", ex.getMessage());
                    }
                });
    }
}

File: backend/storage-service/src/main/kotlin/com/example/storageservice/kafka/producer/KafkaProducerService.kt
package com.example.storageservice.kafka.producer

import org.apache.kafka.clients.producer.ProducerRecord
import org.slf4j.LoggerFactory
import org.springframework.kafka.core.KafkaTemplate
import org.springframework.stereotype.Service

@Service
class KafkaProducerService(private val kafkaTemplate: KafkaTemplate<String, Any>) {

    private val log = LoggerFactory.getLogger(KafkaProducerService::class.java)

    fun sendMessage(key: String?, message: Any) {
        // 키가 null 일 수 있어(옛 자바와 같다) ProducerRecord 로 보낸다.
        kafkaTemplate.send(ProducerRecord(TOPIC, key, message)).whenComplete { result, ex ->
            if (ex == null) {
                log.info("Message sent successfully: {}", result.recordMetadata.topic() + " / " + message)
            } else {
                log.info("Message sent failed: {}", ex.message)
            }
        }
    }

    companion object {
        private const val TOPIC = "topic-member-rollback"
    }
}
